add unbacked strict mode

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147333

fixes #145775

This is the first step in introducing a "strict" mode where we don't silent specialize and don't silent graph break. At a high level when we do mark_unbacked(... strict=True), anytime we specialize an unbacked symint we will explicitly error and tell the user their unbacked dimension was specialized to a single value.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
add unbacked strict mode

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add unbacked strict mode"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add unbacked strict mode"


fixes #145775

[ghstack-poisoned]

laithsakka commented on torch/_dynamo/decorators.py:
```diff
@@ -437,19 +437,28 @@ class directly; instead, use :func:`mark_dynamic`.
 
 @forbid_in_graph
-def mark_unbacked(t, index):
+def mark_unbacked(t, index, strict=False):
```
do you want to make strict=True default in a later diff? or do you want to make mark_dynamic(strict = False) instead?

laithsakka commented on torch/_dynamo/decorators.py:
```diff
@@ -437,19 +437,28 @@ class directly; instead, use :func:`mark_dynamic`.
     raise an exception.  In other words, we will treat this dimension as if it was
     data dependent (we do not know anything about its value.)
+
+    For historical reasons, by default if an unbacked dim is specialized, we will
```
how much work does it take to break the history, i think alternativly we can make this consistent with mark_dynamic by making strict=False default for mark dynamic?

bobrenjc93 replied:
I think in general breaking backwards compatibility is a big deal. We could explore doing it internally via a just knob though.

laithsakka commented on torch/_dynamo/decorators.py:
```diff
@@ -437,19 +437,28 @@ class directly; instead, use :func:`mark_dynamic`.
+                t._dynamo_strict_unbacked_indices = set()
+            t._dynamo_strict_unbacked_indices.add(index)
+
         if not hasattr(t, "_dynamo_unbacked_indices"):
```
do we need else

bobrenjc93 replied:
I don't think so?

laithsakka replied:
ah you want to add both in _dynamo_strict_unbacked_indices and _dynamo_unbacked_indices?

bobrenjc93 replied:
Yeah the alternative is to change it so instead of a set of ints, we have a set of objects, but I figured it's cleaner to stay consistent with the other `indices` and keep the type as a list of sets of ints

laithsakka commented on torch/_dynamo/variables/builder.py:
```diff
@@ -2804,6 +2809,8 @@ def update_dim2constraint(dim, constraint_range, name):
 
         if marked_unbacked:
             dynamic_size = DimDynamic.SIZE_LIKE_UNBACKED
```
why did you need to introduce STRICT_SIZE_LIKE_UNBACKED?

bobrenjc93 replied:
To preserve backwards compatibility when non strict.

bobrenjc93 replied:
Actually on mobile I didn't see your comment. So marked_strict_unbacked is actually derived from STRICT_SIZE_LIKE_UNBACKED

laithsakka replied:
mm I am not sure I understand
```
 marked_strict_unbacked = i in getattr(
            e, "_dynamo_strict_unbacked_indices", set()
        )
 ```
 ...
 ```
 elif marked_strict_unbacked:
                constraint_size = RelaxedUnspecConstraint(warn_only=False)
 ```

laithsakka replied:
we never read             DimDynamic.STRICT_SIZE_LIKE_UNBACKED, except at line 4421 ored with SIZE_LIKE_UNBACKED sounds not needed.

also we do not have DYNAMIC and STRICT_DYNAMIC, so lets be consistent with both?

bobrenjc93 replied:
Sorry I misunderstood, I'll remove it now

laithsakka reviewed (commented):
left some comments can you add a summary?

bobrenjc93 committed:
Update on "add unbacked strict mode"


fixes #145775

This is the first step in introducing a "strict" mode where we don't silent specialize and don't silent graph break. At a high level when we do mark_unbacked(... strict=True), anytime we specialize an unbacked symint we will explicitly error and tell the user their unbacked dimension was specialized to a single value.

[ghstack-poisoned]

laithsakka reviewed (approved):
just one comment

bobrenjc93 committed:
Update on "add unbacked strict mode"


fixes #145775

This is the first step in introducing a "strict" mode where we don't silent specialize and don't silent graph break. At a high level when we do mark_unbacked(... strict=True), anytime we specialize an unbacked symint we will explicitly error and tell the user their unbacked dimension was specialized to a single value.

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 commented:
@pytorchbot merge -i
